lib/lower/convert-type.cc

```diff
@@ -0,0 +1,492 @@
+//===-- lib/lower/convert-type.cc -----------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "convert-type.h"
+#include "../semantics/expression.h"
+#include "../semantics/tools.h"
+#include "../semantics/type.h"
+#include "bridge.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/StandardTypes.h"
+#include "optimizer/FIRType.h"
+
+namespace Br = Fortran::lower;
+namespace Co = Fortran::common;
+namespace Ev = Fortran::evaluate;
+namespace M = mlir;
+namespace Pa = Fortran::parser;
+namespace Se = Fortran::semantics;
+
+using namespace Fortran;
+using namespace Fortran::lower;
+
+namespace {
+
+template <typename A>
+bool isConstant(const Ev::Expr<A> &e) {
+  return Ev::IsConstantExpr(SomeExpr{e});
+}
+
+template <typename A>
+int64_t toConstant(const Ev::Expr<A> &e) {
+  auto opt = Ev::ToInt64(e);
+  assert(opt.has_value() && "expression didn't resolve to a constant");
+  return opt.value();
+}
+
+#undef TODO
+#define TODO()                                                                 \
```

Comment:
is there a reason for the back-slash \?

Reply:
It's a C preprocessor line continuation.  `TODO()` expands to the following two lines.